checks.d/mysql.py

```diff
@@ -394,3 +760,516 @@ def _get_server_pid(self, db):
 
         return pid
+
+    def _get_cpumem_from_ps(self, pid):
+        cpu = None
+        mem = None
+        try:
+            if not Platform.is_windows():
+                ps, _, _ = get_subprocess_output(['ps', '-p{0}'.format(pid), '-o', '%cpu,%mem'], self.log)
+                pslines = ps.strip().splitlines()
+                # First line is header, second line is mysql pid
+                if len(pslines) == 2:
+                    fields = pslines[1].split()
+                    cpu = int(fields[0])
+                    mem = int(fields[1])
+        except Exception:
+            self.log.exception("Error while fetching cpu and memory usage from ps.")
+
+        return cpu, mem
+
+    def _get_stats_from_status(self, db):
+        cursor = db.cursor()
+        cursor.execute("SHOW /*!50002 GLOBAL */ STATUS;")
+        results = dict(cursor.fetchall())
+        cursor.close()
+        del cursor
+        return results
+
+    def _get_stats_from_variables(self, db):
+        cursor = db.cursor()
+        cursor.execute("SHOW GLOBAL VARIABLES;")
+        results = dict(cursor.fetchall())
+        cursor.close()
+        del cursor
+        return results
+
+    def _get_binary_log_stats(self, db):
+        cursor = db.cursor()
+        cursor.execute("SHOW MASTER LOGS;")
+        master_logs = dict(cursor.fetchall())
+
+        cursor.close()
+        del cursor
+
+        binary_log_space = 0
+        for key, value in master_logs.iteritems():
+            binary_log_space += value
+
+        return binary_log_space
+
+    def _is_innodb_engine_enabled(self, db):
+        # Whether InnoDB engine is available or not can be found out either
+        # from the output of SHOW ENGINES or from information_schema.ENGINES
+        # table. Later is choosen because that involves no string parsing.
+        cursor = db.cursor()
+        cursor.execute(
+            "select engine from information_schema.ENGINES where engine='InnoDB'")
+
+        return_val = True if cursor.rowcount > 0 else False
+
+        cursor.close()
+        del cursor
+
+        return return_val
+
+    def _get_replica_stats(self, db):
+        cursor = db.cursor()
+        cursor.execute(
+            "SHOW SLAVE STATUS;")
+
+        replica_results = dict(cursor.fetchall())
+        cursor.close()
+        del cursor
+
+        return replica_results
+
+    def _get_stats_from_innodb_status(self, db):
+        # There are a number of important InnoDB metrics that are reported in
+        # InnoDB status but are not otherwise present as part of the STATUS
+        # variables in MySQL. Majority of these metrics are reported though
+        # as a part of STATUS variables in Percona Server and MariaDB.
+        cursor = db.cursor()
+        cursor.execute("SHOW /*!50000 ENGINE*/ INNODB STATUS")
+
+        innodb_status = cursor.fetchone()
+        innodb_status_text = innodb_status[2]
+
+        cursor.close()
+        del cursor
+
+        results = {
```

Comment:
Should we use a `defaultdict(int)` here ? Or do you actually need to whitelist the keys ?